### PR TITLE
Refactor pipelines builder, fix some issues

### PR DESCRIPTION
**Pull request by bogdandrutu**
* Unconfigured receivers are not identified, this was not a real problem in final binaries since the validation of the config catch this.
* Allow configurations to contain "unused" receivers. Receivers that are configured but not used in any pipeline, this was possible already for exporters and processors.
* Remove the enforcement/check that Receiver factories create the same instance for the same config.

Signed-off-by: Bogdan Drutu <bogdandrutu@gmail.com>

### CODE REVIEW

File: service/internal/builder/pipelines.go [hunk 1]
```diff
@@ -0,0 +1,550 @@
...
+
+type builtPipeline struct {
+	lastConsumer baseConsumer
+
+	receivers  []builtComponent
+	processors []builtComponent
+	exporters  []builtComponent
+}
+
+// BuiltPipelines is a map of allExporters created from exporter configs.
```
Comment (fatsheep9146) [inline service/internal/builder/pipelines.go:59]: The comments seem inadequate, only talk about ```allExporters```

File: service/internal/builder/pipelines.go [hunk 2]
```diff
@@ -0,0 +1,550 @@
...
+	pipelineID config.ComponentID,
+) (component.Exporter, error) {
+	cfg, existsCfg := cfgs[id]
+	if !existsCfg {
+		return nil, fmt.Errorf("exporter %q is not configured", id)
+	}
+
+	factory, existsFactory := factories[id.Type()]
+	if !existsFactory {
+		return nil, fmt.Errorf("exporter factory not available for type: %q", id)
```
Comment (fatsheep9146) [inline service/internal/builder/pipelines.go:302]: I think `id` should be `id.Type()`

File: service/internal/builder/pipelines.go [hunk 3]
```diff
@@ -0,0 +1,550 @@
...
+	nexts []baseConsumer,
+) (component.Receiver, error) {
+	cfg, existsCfg := cfgs[id]
+	if !existsCfg {
+		return nil, fmt.Errorf("receiver %q is not configured", id)
+	}
+
+	factory, existsFactory := factories[id.Type()]
+	if !existsFactory {
+		return nil, fmt.Errorf("processor factory not available for type: %q", id)
```
Comment (fatsheep9146) [inline service/internal/builder/pipelines.go:444]: processor factory -> receiver factory

File: service/internal/builder/pipelines.go [hunk 4]
```diff
@@ -0,0 +1,550 @@
...
+		TelemetrySettings: settings,
+		BuildInfo:         buildInfo,
+	}
+	set.TelemetrySettings.Logger = processorLogger(settings.Logger, id, pipelineID)
+
+	proc, err := createProcessor(ctx, set, procCfg, id, pipelineID, next, factory)
+	if err != nil {
+		// Could not create because this processor does not support this data type.
+		if errors.Is(err, component.ErrDataTypeIsNotSupported) {
+			return nil, fmt.Errorf("pipeline %q has a prorcessor %q, which does not support that data type", pipelineID, id)
```
Comment (fatsheep9146) [inline service/internal/builder/pipelines.go:400]: prorcessor -> processor

**Comment by bogdandrutu** [comment]
@mx-psi converting for the moment to draft, since this PR is too large, will send smaller PRs tomorrow

**Comment by mx-psi** [comment]
👍 Ping me when the PRs are ready to review

File: service/internal/builder/pipelines.go (hunk 2 above)
Comment (bogdandrutu) [reply service/internal/builder/pipelines.go]: Done.

File: service/internal/builder/pipelines.go (hunk 3 above)
Comment (bogdandrutu) [reply service/internal/builder/pipelines.go]: Done.

File: service/internal/builder/pipelines.go (hunk 4 above)
Comment (bogdandrutu) [reply service/internal/builder/pipelines.go]: Done.

**Review by mx-psi — APPROVED** [review]
LGTM, only a few nits for consistency

File: CHANGELOG.md [hunk 5]
```diff
@@ -42,7 +42,10 @@
 
 - Update sum field of exponential histograms to make it optional (#5530)
 - Remove redundant extension shutdown call (#5532)
-
+- Refactor pipelines builder, fix some issues (#5512)
+  - Unconfigured receivers are not identified, this was not a real problem in final binaries since the validation of the config catch this.
+  - Allow configurations to contain "unuse" receivers. Receivers that are configured but not used in any pipeline, this was possible already for exporters and processors.
```
[the remark this answers is not in the document]
Comment (bogdandrutu) [reply CHANGELOG.md]: Done.

File: service/internal/pipelines/pipelines.go [hunk 6]
```diff
@@ -134,27 +178,364 @@ func (bps *Pipelines) HandleZPages(w http.ResponseWriter, r *http.Request) {
...
+	case config.TracesDataType:
+		return factory.CreateTracesExporter(ctx, set, cfg)
+
+	case config.MetricsDataType:
+		return factory.CreateMetricsExporter(ctx, set, cfg)
+
+	case config.LogsDataType:
+		return factory.CreateLogsExporter(ctx, set, cfg)
+	}
+	return nil, fmt.Errorf("error creating exporter %q in pipeline %q, data type %s is not supported", id, pipelineID, pipelineID.Type())
```
[the remark this answers is not in the document]
Comment (bogdandrutu) [reply service/internal/pipelines/pipelines.go]: Done.

File: service/internal/pipelines/pipelines.go [hunk 7]
```diff
@@ -134,27 +178,364 @@ func (bps *Pipelines) HandleZPages(w http.ResponseWriter, r *http.Request) {
...
+	}
+	set.TelemetrySettings.Logger = exporterLogger(settings.Logger, id, pipelineID.Type())
+
+	exp, err := createExporter(ctx, set, cfg, id, pipelineID, factory)
+	if err != nil {
+		if errors.Is(err, component.ErrDataTypeIsNotSupported) {
+			// Could not create because this exporter does not support this data type.
+			return nil, fmt.Errorf("pipeline %q has an exporter %q, which does not support that data type", pipelineID, id)
+		}
+		return nil, fmt.Errorf("error creating %v exporter: %w", id, err)
```
[the remark this answers is not in the document]
Comment (bogdandrutu) [reply service/internal/pipelines/pipelines.go]: Done.

**Pull request merged**